Source/WebCore/dom/Element.cpp

```diff
@@ -4740,6 +4743,48 @@ bool Element::isSpellCheckingEnabled() const
     return true;
 }
 
+bool Element::isWritingSuggestionsEnabled() const
+{
+    // If none of the following conditions are true, then return `false`.
+
+    // `element` is an `input` element whose `type` attribute is in either the
+    // `Text`, `Search`, `URL`, `Email` state and is `mutable`.
+    auto isEligibleInputElement = [&] {
+        RefPtr input = dynamicDowncast<HTMLInputElement>(*this);
+        if (!input)
+            return false;
+
+        return !input->isDisabledFormControl() && input->supportsWritingSuggestions();
+    }();
+
+    // `element` is a `textarea` element that is `mutable`.
+    auto isTextArea = is<HTMLTextAreaElement>(*this);
+
+    // `element` is an `editing host` or is `editable`.
+    auto isEditable = hasEditableStyle() || isContentEditable();
```

Comment:
It's unclear why we need to check both bits here. I would recommend just checking `hasEditableStyle()`, and moving this flag into `EditorState::PostLayoutData`, along with the other bits of information on `EditorState` that already rely on layout being up-to-date.

Reply:
> moving this flag
which flag is this referring to? this entire method `isWritingSuggestionsEnabled`?

Reply:
It's referring to the flag you have on `EditorState`.